File: packages/wasmvm/wasmclient/go/wasmclient/wasmclientsandbox.go
// Copyright 2020 IOTA Stiftung
// SPDX-License-Identifier: Apache-2.0

package wasmclient

import (
	"errors"
	"fmt"

	"github.com/iotaledger/wasp/packages/wasmvm/wasmlib/go/wasmlib"
	"github.com/iotaledger/wasp/packages/wasmvm/wasmlib/go/wasmlib/wasmrequests"
	"github.com/iotaledger/wasp/packages/wasmvm/wasmlib/go/wasmlib/wasmtypes"
)

func (s *WasmClientContext) FnCall(req *wasmrequests.CallRequest) []byte {
	if req.Contract != s.scHname {
		s.Err = fmt.Errorf("unknown contract: %s", req.Contract.String())
		return nil
	}

	res, err := s.svcClient.CallViewByHname(req.Contract, req.Function, req.Params)
	if err != nil {
		s.Err = err
		return nil
	}
	return res
}

func (s *WasmClientContext) FnChainID() wasmtypes.ScChainID {
	return s.CurrentChainID()
}

func (s *WasmClientContext) FnPost(req *wasmrequests.PostRequest) []byte {
	if s.keyPair == nil {
		s.Err = errors.New("missing key pair")
		return nil
	}

	if req.ChainID != s.CurrentChainID() {
		s.Err = fmt.Errorf("unknown chain id: %s", req.ChainID.String())
		return nil
	}

	if req.Contract != s.scHname {
		s.Err = fmt.Errorf("unknown contract: %s", req.Contract.String())
		return nil
	}

	scAssets := wasmlib.NewScAssets(req.Transfer)
	s.ReqID, s.Err = s.svcClient.PostRequest(req.ChainID, req.Contract, req.Function, req.Params, scAssets, s.keyPair)
	return nil
}

File: packages/wasmvm/wasmclient/ts/wasmclient/lib/wasmclientsandbox.ts
// Copyright 2020 IOTA Stiftung
// SPDX-License-Identifier: Apache-2.0

import * as isc from './isc';
import * as wasmlib from 'wasmlib';
import {WasmClientService} from './';

export class WasmClientSandbox {
    Err: isc.Error = null;
    keyPair: isc.KeyPair | null = null;
    nonce: u64 = 0n;
    ReqID: wasmlib.ScRequestID = new wasmlib.ScRequestID();
    scName: string;
    scHname: wasmlib.ScHname;
    svcClient: WasmClientService;

    public constructor(svcClient: WasmClientService, scName: string) {
        this.svcClient = svcClient;
        this.scName = scName;
        this.scHname = wasmlib.hnameFromBytes(isc.Codec.hNameBytes(scName));
    }

    public currentChainID(): wasmlib.ScChainID {
        return this.svcClient.currentChainID();
    }

    public fnCall(req: wasmlib.CallRequest): Uint8Array {
        if (!req.contract.equals(this.scHname)) {
            this.Err = 'unknown contract: ' + req.contract.toString();
            return new Uint8Array(0);
        }

        const [res, err] = this.svcClient.callViewByHname(req.contract, req.function, req.params);
        this.Err = err;
        return res;
    }

    public fnChainID(): wasmlib.ScChainID {
        return this.currentChainID();
    }

    public fnPost(req: wasmlib.PostRequest): Uint8Array {
        if (this.keyPair == null) {
            this.Err = 'missing key pair';
            return new Uint8Array(0);
        }

        if (!req.chainID.equals(this.currentChainID())) {
            this.Err = 'unknown chain id: ' + req.chainID.toString();
            return new Uint8Array(0);
        }

        if (!req.contract.equals(this.scHname)) {
            this.Err = 'unknown contract:' + req.contract.toString();
            return new Uint8Array(0);
        }

        const scAssets = new wasmlib.ScAssets(req.transfer);
        [this.ReqID, this.Err] = this.svcClient.postRequest(req.chainID, req.contract, req.function, req.params, scAssets, this.keyPair);
        return new Uint8Array(0);
    }
}
